### PR TITLE
ARROW-849: [C++] Support setting production build dependencies with ARROW_BUILD_TOOLCHAIN

### PR DESCRIPTION
**Pull request by wesm**
Opening this for comment. If we like this, we can do the same thing in parquet-cpp. Will need to be documented in the README. 

I did not use the environment variable for gflags/gtest/gbenchmark, since this are test/benchmark-only dependencies, and they build automatically when needed.

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -382,6 +375,40 @@ enable_testing()
...
+  else()
+    set(RAPIDJSON_HOME "$ENV{RAPIDJSON_HOME}")
+  endif()
+
+  if ("$ENV{JEMALLOC_HOME}" STREQUAL "")
+    set(JEMALLOC_HOME "$ENV{ARROW_BUILD_TOOLCHAIN}")
+  else()
+    set(JEMALLOC_HOME "$ENV{JEMALLOC_HOME}")
+  endif()
+endif()
```
Comment (wesm) [inline cpp/CMakeLists.txt]: Argh, I realized this does not handle the case where this environment variable is not set. Will revise

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -382,6 +375,38 @@ enable_testing()
...
+set(GTEST_VERSION "1.8.0")
+set(GBENCHMARK_VERSION "1.1.0")
+set(FLATBUFFERS_VERSION "1.6.0")
+set(JEMALLOC_VERSION "4.4.0")
+
+set(FLATBUFFERS_HOME "$ENV{FLATBUFFERS_HOME}")
+set(RAPIDJSON_HOME "$ENV{RAPIDJSON_HOME}")
+set(JEMALLOC_HOME "$ENV{JEMALLOC_HOME}")
+
+if (NOT "$ENV{ARROW_BUILD_TOOLCHAIN}" STREQUAL "")
```
Comment (xhochy) [inline cpp/CMakeLists.txt]: I would turn the logic around here. If `ARROW_BUILD_TOOLCHAIN` is set, use it. But if a more specific environment variable like `JEMALLOC_HOME` is set, it should override `ARROW_BUILD_TOOLCHAIN`. Thus you can use e.g. conda for all except `jemalloc` which you want to use the system provided one.

**Review by xhochy — CHANGES_REQUESTED** [review]

**Comment by wesm** [comment]
Done. Now you can override individual toolchain variables (including unsetting them / reverting to built-in thirdparty) by setting:

```
export ARROW_BUILD_TOOLCHAIN=$HOME/my-toolchain
export FLATBUFFERS_HOME=
```

**Comment by wesm** [comment]
Also updated https://github.com/apache/parquet-cpp/pull/304 accordingly

**Comment by wesm** [comment]
@xhochy this look ok?

**Comment by wesm** [comment]
+1. I will update the conda-forge recipes to use this in both projects

**Pull request closed without merging**